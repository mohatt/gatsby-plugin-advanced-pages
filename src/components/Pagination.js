import React from 'react'
import { shape, number, string, object, bool, element, oneOf, oneOfType } from 'prop-types'
import classNames from 'classnames'
import Link from './Link'

export default class Pagination extends React.Component {
  static propTypes = {
    route: string.isRequired,
    params: object,
    ui: oneOf(['mini', 'simple', 'full']),
    range: number,
    className: string,
    pageInfo: shape({
      itemCount: number.isRequired,
      perPage: number.isRequired,
      pageCount: number.isRequired,
      currentPage: number.isRequired,
      hasNextPage: bool.isRequired,
      hasPreviousPage: bool.isRequired
    }).isRequired,
    labels: shape({
      prev: oneOfType([string, element]),
      next: oneOfType([string, element]),
      first: oneOfType([string, element]),
      last: oneOfType([string, element])
    }),
    theme: shape({
      inner: string,
      item: string,
      'item.next': string,
      'item.prev': string,
      'item.first': string,
      'item.last': string,
      link: string,
      active: string,
      disabled: string
    }),
    renderDisabled: bool
  };

  static defaultProps = {
    ui: 'full',
    range: 6,
    labels: {
      first: '« First',
      prev: '← Previous',
      next: 'Next →',
      last: 'Last »'
    },
    theme: {
      inner: 'pagination',
      item: 'page-item',
      'item.next': 'page-item__next',
      'item.prev': 'page-item__prev',
      'item.first': 'page-item__first',
      'item.last': 'page-item__last',
      link: 'page-link',
      active: 'active',
      disabled: 'disabled'
    },
    renderDisabled: true
  }

  render () {
    const {
      ui,
      pageInfo: {
        pageCount,
        currentPage,
        hasNextPage,
        hasPreviousPage
      }
    } = this.props

    const labels = Object.assign({}, Pagination.defaultProps.labels, this.props.labels)
    const theme = Object.assign({}, Pagination.defaultProps.theme, this.props.theme)
    const pages = []

    if (ui === 'full') {
      pages.push({
        key: 'first',
        number: 1,
        type: 'first',
        label: labels.first,
        disabled: !hasPreviousPage
      })
    }

    pages.push({
      key: 'prev',
      number: currentPage - 1,
      type: 'prev',
      label: labels.prev,
      disabled: !hasPreviousPage
    })

    if (ui !== 'mini') {
      const [fp, lp] = this.calcRange()
      for (let i = fp; i <= lp; i++) {
        pages.push({
          key: i,
          number: i,
          type: 'page',
          label: i.toString(),
          active: i === currentPage
        })
      }
    }

    pages.push({
      key: 'next',
      number: currentPage + 1,
      type: 'next',
      label: labels.next,
      disabled: !hasNextPage
    })

    if (ui === 'full') {
      pages.push({
        key: 'last',
        number: pageCount,
        type: 'last',
        label: labels.last,
        disabled: !hasNextPage
      })
    }

    return (
      <nav className={this.props.className} role='navigation' aria-label='Pagination Navigation'>
        <ul className={theme.inner}>
          {pages.map((page) => {
            if (page.disabled && !this.props.renderDisabled) {
              return null
            }

            const classes = [theme.item]
            const typeClass = theme['item.' + page.type]
            if (typeClass) {
              classes.push(typeClass)
            }

            classes.push({
              [theme.active]: page.active,
              [theme.disabled]: page.disabled
            })

            if (page.disabled) {
              return (
                <li key={page.key} className={classNames(classes)}>
                  <a className={theme.link}>{page.label}</a>
                </li>
              )
            }

            // Use non-paginated route for first page
            const [params, scope] = page.number === 1
              ? [this.props.params, null]
              : [{ ...this.props.params, page: page.number }, 'pagination']

            return (
              <li key={page.key} className={classNames(classes)}>
                <Link to={this.props.route} params={params} scope={scope} className={theme.link}>
                  {page.label}
                </Link>
              </li>
            )
          })}
        </ul>
      </nav>
    )
  }

  calcRange () {
    const { range, pageInfo: { currentPage, pageCount } } = this.props

    let fp = Math.max(1, currentPage - Math.floor(range / 2))
    let lp = Math.min(pageCount, currentPage + Math.floor(range / 2))

    if (lp - fp + 1 < range) {
      if (currentPage < pageCount / 2) {
        lp = Math.min(
          pageCount,
          lp + (range - (lp - fp))
        )
      } else {
        fp = Math.max(1, fp - (range - (lp - fp)))
      }
    }

    if (lp - fp + 1 > range) {
      if (currentPage > pageCount / 2) {
        fp = fp + 1
      } else {
        lp = lp - 1
      }
    }

    return [fp, lp]
  }
}
