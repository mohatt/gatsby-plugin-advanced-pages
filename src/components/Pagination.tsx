import React, { Component, ReactNode } from 'react'
import { shape, number, string, object, bool, element, oneOf, oneOfType } from 'prop-types'
import { clsx, ClassArray } from 'clsx'
import { Link } from './Link'
import type { RouteParams } from '../lib/route-compiler'

export interface PaginationInfo {
  itemCount: number
  perPage: number
  pageCount: number
  currentPage: number
  hasNextPage: boolean
  hasPreviousPage: boolean
}

export interface PaginationProps {
  route: string
  params?: RouteParams
  ui?: 'mini' | 'simple' | 'full'
  range?: number
  className?: string
  renderDisabled?: boolean
  labels?: {
    prev?: ReactNode
    next?: ReactNode
    first?: ReactNode
    last?: ReactNode
  }
  theme?: {
    inner?: string
    item?: string
    'item.next'?: string
    'item.prev'?: string
    'item.first'?: string
    'item.last'?: string
    link?: string
    active?: string
    disabled?: string
  }
  pageInfo: PaginationInfo
}

export class Pagination extends Component<PaginationProps> {
  static propTypes = {
    route: string.isRequired,
    params: object,
    ui: oneOf(['mini', 'simple', 'full']),
    range: number,
    className: string,
    pageInfo: shape({
      itemCount: number.isRequired,
      perPage: number.isRequired,
      pageCount: number.isRequired,
      currentPage: number.isRequired,
      hasNextPage: bool.isRequired,
      hasPreviousPage: bool.isRequired
    }).isRequired,
    labels: shape({
      prev: oneOfType([string, element]),
      next: oneOfType([string, element]),
      first: oneOfType([string, element]),
      last: oneOfType([string, element])
    }),
    theme: shape({
      inner: string,
      item: string,
      'item.next': string,
      'item.prev': string,
      'item.first': string,
      'item.last': string,
      link: string,
      active: string,
      disabled: string
    }),
    renderDisabled: bool
  }

  static defaultProps: Omit<PaginationProps, 'route' | 'pageInfo'> = {
    ui: 'full',
    range: 6,
    labels: {
      first: '« First',
      prev: '← Previous',
      next: 'Next →',
      last: 'Last »'
    },
    theme: {
      inner: 'pagination',
      item: 'page-item',
      'item.next': 'page-item__next',
      'item.prev': 'page-item__prev',
      'item.first': 'page-item__first',
      'item.last': 'page-item__last',
      link: 'page-link',
      active: 'active',
      disabled: 'disabled'
    },
    renderDisabled: true
  }

  render () {
    // Assign default prop values
    const props: Readonly<PaginationProps> = {
      ...Pagination.defaultProps,
      ...this.props,
      labels: {
        ...Pagination.defaultProps.labels,
        ...this.props.labels
      },
      theme: {
        ...Pagination.defaultProps.theme,
        ...this.props.theme
      }
    }
    const {
      ui,
      labels,
      theme,
      pageInfo: {
        pageCount,
        currentPage,
        hasNextPage,
        hasPreviousPage
      }
    } = props

    const pages: Array<{
      key: string | number
      number: number
      type: 'first' | 'prev' | 'page' | 'next' | 'last'
      label: ReactNode
      active?: boolean
      disabled?: boolean
    }> = []

    if (ui === 'full') {
      pages.push({
        key: 'first',
        number: 1,
        type: 'first',
        label: labels.first,
        disabled: !hasPreviousPage
      })
    }

    pages.push({
      key: 'prev',
      number: currentPage - 1,
      type: 'prev',
      label: labels.prev,
      disabled: !hasPreviousPage
    })

    if (ui !== 'mini') {
      const [fp, lp] = this.calcRange(pageCount, currentPage, props.range)
      for (let i = fp; i <= lp; i++) {
        pages.push({
          key: i,
          number: i,
          type: 'page',
          label: i.toString(),
          active: i === currentPage
        })
      }
    }

    pages.push({
      key: 'next',
      number: currentPage + 1,
      type: 'next',
      label: labels.next,
      disabled: !hasNextPage
    })

    if (ui === 'full') {
      pages.push({
        key: 'last',
        number: pageCount,
        type: 'last',
        label: labels.last,
        disabled: !hasNextPage
      })
    }

    return (
      <nav className={props.className} role='navigation' aria-label='Pagination Navigation'>
        <ul className={theme.inner}>
          {pages.map((page) => {
            if (page.disabled && !props.renderDisabled) {
              return null
            }

            const classes: ClassArray = [theme.item]
            const typeClass = theme['item.' + page.type]
            if (typeClass) {
              classes.push(typeClass)
            }

            classes.push({
              [theme.active]: page.active,
              [theme.disabled]: page.disabled
            })

            if (page.disabled) {
              return (
                <li key={page.key} className={clsx(classes)}>
                  <a className={theme.link}>{page.label}</a>
                </li>
              )
            }

            // Use non-paginated route for first page
            const [params, scope] = page.number === 1
              ? [props.params, null]
              : [{ ...props.params, page: page.number }, 'pagination'] as const

            return (
              <li key={page.key} className={clsx(classes)}>
                <Link to={props.route} params={params} scope={scope} className={theme.link}>
                  {page.label}
                </Link>
              </li>
            )
          })}
        </ul>
      </nav>
    )
  }

  private calcRange (pageCount: number, currentPage: number, range: number) {
    let fp = Math.max(1, currentPage - Math.floor(range / 2))
    let lp = Math.min(pageCount, currentPage + Math.floor(range / 2))

    if (lp - fp + 1 < range) {
      if (currentPage < pageCount / 2) {
        lp = Math.min(
          pageCount,
          lp + (range - (lp - fp))
        )
      } else {
        fp = Math.max(1, fp - (range - (lp - fp)))
      }
    }

    if (lp - fp + 1 > range) {
      if (currentPage > pageCount / 2) {
        fp = fp + 1
      } else {
        lp = lp - 1
      }
    }

    return [fp, lp]
  }
}

export default Pagination
