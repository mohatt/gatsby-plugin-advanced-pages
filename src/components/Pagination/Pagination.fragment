import { graphql } from "gatsby"

export const pagination = graphql`
  fragment Pagination on PageInfo {
    perPage
    pageCount
    itemCount
    currentPage
    hasNextPage
    hasPreviousPage
  }
`
