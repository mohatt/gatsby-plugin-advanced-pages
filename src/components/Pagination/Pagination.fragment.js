import { graphql } from 'gatsby'

export const Pagination = graphql`
  fragment Pagination on PageInfo {
    perPage
    pageCount
    itemCount
    currentPage
    hasNextPage
    hasPreviousPage
  }
`
