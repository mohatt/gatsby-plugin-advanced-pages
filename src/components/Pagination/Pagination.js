import React from 'react'
import { shape, number, string, object, bool, element, oneOf, oneOfType } from 'prop-types'
import classNames from 'classnames'
import Link from '../Link'
import { getPageInfo } from 'paging-algorithm'

export default class Pagination extends React.Component {
  static propTypes = {
    route: string.isRequired,
    params: object,
    ui: oneOf(['mini', 'simple', 'full']),
    range: number,
    className: string,
    pageInfo: shape({
      itemCount: number.isRequired,
      currentPage: number.isRequired,
      perPage: number.isRequired
    }).isRequired,
    labels: shape({
      prev: oneOfType([string, element]),
      next: oneOfType([string, element]),
      first: oneOfType([string, element]),
      last: oneOfType([string, element])
    }),
    theme: shape({
      inner: string,
      item: string,
      'item.next': string,
      'item.prev': string,
      'item.first': string,
      'item.last': string,
      link: string,
      active: string,
      disabled: string
    }),
    renderDisabled: bool
  };

  static defaultProps = {
    ui: 'full',
    range: 6,
    labels: {
      first: '« First',
      prev: '⟨ Previous',
      next: 'Next ⟩',
      last: 'Last »'
    },
    theme: {
      inner: 'pagination',
      item: 'page-item',
      'item.next': 'page-item__next',
      'item.prev': 'page-item__prev',
      'item.first': 'page-item__first',
      'item.last': 'page-item__last',
      link: 'page-link',
      active: 'active',
      disabled: 'disabled'
    },
    renderDisabled: true
  }

  constructor(props) {
    super(props)

    this.state = getPageInfo({
      limit: props.pageInfo.perPage,
      pageCount: props.range,
      total: props.pageInfo.itemCount,
      page: props.pageInfo.currentPage
    })
  }

  render() {
    const {
      totalPages,
      hasPreviousPage,
      hasNextPage,
      previousPage,
      nextPage,
      firstPage,
      lastPage,
      currentPage
    } = this.state

    if(!totalPages || !/^\d+$/.test(totalPages)){
      if(process.env.NODE_ENV !== `production`){
        console.error(
          'Warning: Invalid pageInfo prop supplied to `Pagination`' +
          ' component: ' + JSON.stringify(this.props.pageInfo))
      }
      return <div className={this.props.className}>Pagination not available</div>
    }

    const { ui } = this.props
    const labels = Object.assign({}, Pagination.defaultProps.labels, this.props.labels)
    const theme = Object.assign({}, Pagination.defaultProps.theme, this.props.theme)
    const pages = []

    if(ui === 'full') {
      pages.push({
        key: 'first',
        number: 1,
        type: 'first',
        label: labels.first,
        disabled: !hasPreviousPage
      })
    }

    pages.push({
      key: 'prev',
      number: previousPage,
      type: 'prev',
      label: labels.prev,
      disabled: !hasPreviousPage
    })

    if(ui !== 'mini') {
      for (let i = firstPage; i <= lastPage; i++) {
        pages.push({
          key: i,
          number: i,
          type: 'page',
          label: i.toString(),
          active: i === currentPage
        })
      }
    }

    pages.push({
      key: 'next',
      number: nextPage,
      type: 'next',
      label: labels.next,
      disabled: !hasNextPage
    })

    if(ui === 'full') {
      pages.push({
        key: 'last',
        number: totalPages,
        type: 'last',
        label: labels.last,
        disabled: !hasNextPage
      })
    }

    return (
      <nav className={this.props.className} role="navigation" aria-label="Pagination Navigation">
        <ul className={theme.inner}>
        {pages.map((page) => {
          if(page.disabled && !this.props.renderDisabled){
            return
          }

          const classes = [theme.item]
          const typeClass = theme['item.' + page.type]
          if(typeClass){
            classes.push(typeClass)
          }

          classes.push({
            [theme.active]: page.active,
            [theme.disabled]: page.disabled
          })

          if(page.disabled){
            return (
              <li key={page.key} className={classNames(classes)}>
                <a className={theme.link}>{page.label}</a>
              </li>
            )
          }

          // Use non-paginated route for first page 
          const [ params, scope ] = page.number === 1
            ? [ this.props.params, null ]
            : [ { ...this.props.params, page: page.number }, 'pagination' ]

          return (
            <li key={page.key} className={classNames(classes)}>
              <Link
                to={this.props.route}
                params={params}
                scope={scope}
                className={theme.link}
                children={page.label}
              />
            </li>
          )
        })}
        </ul>
      </nav>
    )
  }
}
