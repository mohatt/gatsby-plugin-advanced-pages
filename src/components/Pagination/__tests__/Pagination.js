import React from "react"
import TestRenderer from 'react-test-renderer';
import Pagination from "../Pagination"

const setup = (props = {}) => {
  return TestRenderer.create(
    <Pagination {...props} />
  ).toJSON()
}

describe(`<Pagination />`, () => {
  it(`should render correctly`, () => {
    const result = setup({
      route: 'blog',
      pageInfo: {
        perPage: 10,
        itemCount: 80,
        currentPage: 3,
      }
    })
    expect(result).toMatchSnapshot()
  })
})
