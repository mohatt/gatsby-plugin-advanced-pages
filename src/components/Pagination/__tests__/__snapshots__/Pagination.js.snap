// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> should render correctly 1`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    className="pagination"
  >
    <li
      className="page-item page-item__first"
    >
      <a
        className="page-link"
        href="/blog/page/1"
      >
        « First
      </a>
    </li>
    <li
      className="page-item page-item__prev"
    >
      <a
        className="page-link"
        href="/blog/page/2"
      >
        ⟨ Previous
      </a>
    </li>
    <li
      className="page-item"
    >
      <a
        className="page-link"
        href="/blog/page/1"
      >
        1
      </a>
    </li>
    <li
      className="page-item"
    >
      <a
        className="page-link"
        href="/blog/page/2"
      >
        2
      </a>
    </li>
    <li
      className="page-item active"
    >
      <a
        className="page-link"
        href="/blog/page/3"
      >
        3
      </a>
    </li>
    <li
      className="page-item"
    >
      <a
        className="page-link"
        href="/blog/page/4"
      >
        4
      </a>
    </li>
    <li
      className="page-item"
    >
      <a
        className="page-link"
        href="/blog/page/5"
      >
        5
      </a>
    </li>
    <li
      className="page-item"
    >
      <a
        className="page-link"
        href="/blog/page/6"
      >
        6
      </a>
    </li>
    <li
      className="page-item page-item__next"
    >
      <a
        className="page-link"
        href="/blog/page/4"
      >
        Next ⟩
      </a>
    </li>
    <li
      className="page-item page-item__last"
    >
      <a
        className="page-link"
        href="/blog/page/8"
      >
        Last »
      </a>
    </li>
  </ul>
</nav>
`;
